WcaOnRails/spec/controllers/api_competitions_controller_spec.rb

```diff
@@ -40,4 +40,279 @@
+        json = JSON.parse(response.body)
+        expect(json.length).to eq 1
+        expect(json[0]["id"]).to eq competition.id
+      end
+    end
+
+    it 'can do a plaintext query' do
+      terrible_comp = FactoryGirl.create(:competition, :confirmed, :visible, name: "A terrible competition 2016", countryId: "USA")
+      awesome_comp = FactoryGirl.create(:competition, :confirmed, :visible, name: "An awesome competition 2016", countryId: "France")
+
+      get :index, params: { q: "AWES" }
+      json = JSON.parse(response.body)
+      expect(json.length).to eq 1
+      expect(json[0]["id"]).to eq awesome_comp.id
+
+      # Check that composing a plaintext query and a country query works.
+      get :index, params: { q: "competition", country_iso2: "US" }
+      json = JSON.parse(response.body)
+      expect(json.length).to eq 1
+      expect(json[0]["id"]).to eq terrible_comp.id
+    end
+
+    it 'validates start' do
+      get :index, params: { start: "2015" }
+      expect(response.status).to eq 422
+      json = JSON.parse(response.body)
+      expect(json["errors"]).to eq ["Invalid start: '2015'"]
+    end
+
+    it 'validates end' do
+      get :index, params: { end: "2014" }
+      expect(response.status).to eq 422
+      json = JSON.parse(response.body)
+      expect(json["errors"]).to eq ["Invalid end: '2014'"]
+    end
+
+    it 'validates country_iso2' do
+      get :index, params: { country_iso2: "this is not a country" }
+      expect(response.status).to eq 422
+      json = JSON.parse(response.body)
+      expect(json["errors"]).to eq ["Invalid country_iso2: 'this is not a country'"]
+    end
+
+    it 'can query by date' do
+      last_feb_comp = FactoryGirl.create(:competition, :confirmed, :visible, starts: Date.new(2015, 2, 1))
+      feb_comp = FactoryGirl.create(:competition, :confirmed, :visible, starts: Date.new(2016, 2, 1))
+      march_comp = FactoryGirl.create(:competition, :confirmed, :visible, starts: Date.new(2016, 3, 1))
+
+      get :index, params: { start: "2015-02-01" }
+      json = JSON.parse(response.body)
+      expect(json.map { |c| c["id"] }).to eq [march_comp.id, feb_comp.id, last_feb_comp.id]
+
+      get :index, params: { end: "2016-03-01" }
+      json = JSON.parse(response.body)
+      expect(json.map { |c| c["id"] }).to eq [march_comp.id, feb_comp.id, last_feb_comp.id]
+
+      get :index, params: { start: "2015-02-01", end: "2016-02-15" }
+      json = JSON.parse(response.body)
+      expect(json.map { |c| c["id"] }).to eq [feb_comp.id, last_feb_comp.id]
+
+      get :index, params: { start: "2015-02-01", end: "2015-02-01" }
+      json = JSON.parse(response.body)
+      expect(json.map { |c| c["id"] }).to eq [last_feb_comp.id]
+    end
+
+    it 'paginates' do
+      30.times do
+        FactoryGirl.create :competition, :confirmed, :visible
+      end
+
+      get :index
+      expect(response.status).to eq 200
+      json = JSON.parse(response.body)
+      expect(json.length).to eq 25
```

Comment:
Good call! Yes, that 25 does come from ["far away'](https://github.com/davidcelis/api-pagination#rails). I've changed the test to explicitly specify per_page as 5 here, and reduced the number of competitions we create to from 30 to 7.